### PR TITLE
Authentication

### PR DESCRIPTION
**Pull request by adriennedomingus**

### CODE REVIEW

File: app/controllers/sessions_controller.rb [hunk 1]
```diff
@@ -0,0 +1,17 @@
+class SessionsController < ApplicationController
+  def create
+    @user = User.find_by(email: params[:session][:email])
+    if @user.authenticate(params[:session][:password])
+      session[:user_id] = @user.id
+      redirect_to @user
+    else
+      flash.now[:notice] = "Your email or "
```
Comment (adriennedomingus) [inline app/controllers/sessions_controller.rb]: Will go in and fix this. Obviously need a test for invalid login also.

File: app/controllers/sessions_controller.rb [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+      render :new
+    end
+  end
+
+  def destroy
+    session.clear
```
Comment (amaxwellblair) [inline app/controllers/sessions_controller.rb]: Maybe add a flash to confirm log out

File: app/controllers/users_controller.rb [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+  end
+
+  def create
+    @user = User.new(user_params)
+    if @user.save
+      @user.update(api_key: SecureRandom.hex)
+      session[:user_id] = @user.id
+      redirect_to @user
+    else
+      flash.now[:success] = "There was a problem with your request. Please make sure you entered a valid email that has not been used to generate an api key before, and that your passwords match"
```
Comment (amaxwellblair) [inline app/controllers/users_controller.rb]: `flash[:success]` could be changed to `flash[:failure]`

given this seems like a sad path

File: app/controllers/api/v1/demographics/districts_controller.rb [hunk 4]
```diff
@@ -4,13 +4,16 @@ class Api::V1::Demographics::DistrictsController < Api::ApiController
   def show
     district = District.find_by(slug: params["slug"])
     year = SchoolYear.find_by(years: params["year"])
+    key = params[:api_key]
     if invalid_request(district, year)
       message = invalid_request(district, year)
```
Comment (amaxwellblair) [inline app/controllers/api/v1/demographics/districts_controller.rb]: not the most common pattern in RoR but one way you could clean this up is:

```
  if message = invalid_request(district, year); message
        ...
  end
```

Comment (geopet) [inline app/controllers/api/v1/demographics/districts_controller.rb]: Totally agree with @amaxwellblair that you could one line that if you wanted. On the other hand, I tend to take the opposite path. I am more verbose and use more white space and more of everything. I just find that trying to dissect a one liner when something is breaking at 1am is just not worth it to me.

For the record, a lot of people disagree with my point of view on this. 😆

**Pull request merged**